### PR TITLE
Training data crf

### PR DESCRIPTION
**Pull request by jayaraopratik**
This PR adds the following functionalities
1. Create new index with new mappings for training data
2. Create Alias method used to add aliases
3. Add API endpoint and methods to get training data
4. Add API endpoint and methods to update training data

### CODE REVIEW

File: datastore/exceptions.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
     'IndexForTransferException', 'AliasForTransferException', 'NonESEngineTransferException',
     'IndexNotFoundException', 'InvalidESURLException', 'SourceDestinationSimilarException',
     'InternalBackupException', 'AliasNotFoundException', 'PointIndexToAliasException',
-    'FetchIndexForAliasException', 'DeleteIndexFromAliasException'
+    'FetchIndexForAliasException', 'DeleteIndexFromAliasException', 'TrainingIndexNotConfigured'
```
Comment (amansrivastava17) [inline datastore/exceptions.py]: add exception

File: datastore/exceptions.py [hunk 2]
```diff
@@ -187,3 +187,17 @@ def __init__(self, message=None):
 
     def __str__(self):
         return repr(self.value)
+
+
+class TrainingIndexNotConfigured(Exception):
```
Comment (amansrivastava17) [inline datastore/exceptions.py]: Rename same as above

File: external_api/api.py [hunk 3]
```diff
@@ -7,9 +7,10 @@
 from datastore.exceptions import IndexNotFoundException, InvalidESURLException, \
     SourceDestinationSimilarException, \
     InternalBackupException, AliasNotFoundException, PointIndexToAliasException, \
-    FetchIndexForAliasException, DeleteIndexFromAliasException
+    FetchIndexForAliasException, DeleteIndexFromAliasException, TrainingIndexNotConfigured
```
Comment (amansrivastava17) [inline external_api/api.py]: rename

File: datastore/constants.py [hunk 4]
```diff
@@ -13,3 +13,5 @@
 ELASTICSEARCH_INDEX_NAME = 'name'
 ELASTICSEARCH_DOC_TYPE = 'doc_type'
 ELASTICSEARCH_VERSION_MAJOR, ELASTICSEARCH_VERSION_MINOR, ELASTICSEARCH_VERSION_OTHER = elasticsearch.VERSION
+ES_TRAINING_INDEX = 'es_training_index'
```
Comment (amansrivastava17) [inline datastore/constants.py]: Add comment for purpose it is being used

File: datastore/datastore.py [hunk 5]
```diff
@@ -393,3 +405,92 @@ def transfer_entities_elastic_search(self, entity_list):
...
+        results_dictionary = {}
+        if self._engine == ELASTICSEARCH:
+            self._check_doc_type_for_training_data_elasticsearch()
+
+            es_training_index = self._connection_settings.get('es_training_index')
+            if es_training_index is None:
+                raise TrainingIndexNotConfigured()
+
+            request_timeout = self._connection_settings.get('request_timeout', 20)
+            results_dictionary = elastic_search.query.training_data_query(connection=self._client_or_connection,
```
Comment (amansrivastava17) [inline datastore/datastore.py]: rename method `training_data_query` as below mentioned

File: datastore/elastic_search/create.py [hunk 6]
```diff
@@ -119,3 +104,105 @@ def exists(connection, index_name):
...
+        Refer https://elasticsearch-py.readthedocs.io/en/master/api.html#elasticsearch.client.IndicesClient.create
+        Refer https://elasticsearch-py.readthedocs.io/en/master/api.html#elasticsearch.client.IndicesClient.put_mapping
+    """
+    mapping_body = {
+        doc_type: {
+            'properties': {
+                "entity_data": {
+                    "type": "text"
+                },
+                "text": {
```
Comment (amansrivastava17) [inline datastore/elastic_search/create.py]: change `text` to sentence as its look confusing with text type

File: datastore/exceptions.py (hunk 1 above)
Comment (jayaraopratik) [reply datastore/exceptions.py]: Done

**Pull request merged**